### PR TITLE
Update custom storage visualizations feature

### PR DESCRIPTION
**Pull request by neokore**
- `updateVisualization` method implemented.
- ~⚠️ Note that although a visualization is unlinked to a dataset, the related dataset is not deleted.~
- Some column names were changed due to conflicts with reserved names.

Related to https://github.com/CartoDB/toolkit/issues/52

### CODE REVIEW

File: packages/custom-storage/__tests__/SQLStorage.spec.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ describe('SQLStorage', () => {
 
   it('should check or create tables for storage on init', async () => {
     const expectedQueries = [
-      'CREATE TABLE IF NOT EXISTS mynamespace_public_v1 (id uuid PRIMARY KEY DEFAULT mynamespace_create_uuid(), name text NOT NULL, description text, thumbnail text, private boolean, config json, last_modified timestamp NOT NULL DEFAULT now());',
+      'CREATE TABLE IF NOT EXISTS mynamespace_public_v1 (id uuid PRIMARY KEY DEFAULT mynamespace_create_uuid(), name text NOT NULL, description text, thumbnail text, isprivate boolean, config json, lastmodified timestamp NOT NULL DEFAULT now());',
```
Comment (VictorVelarde) [inline packages/custom-storage/__tests__/SQLStorage.spec.ts]: I now about the private 'reserved' word issue, but out of curiosity, why did you change 'last_modified'?

**Comment by VictorVelarde** [comment]
I have not tested it but I think we need a bit of a refactor, cause insert & update share a lot of code. Let me give it a try

**Comment by VictorVelarde** [comment]
I made a small refactor. I hope everything's fine regarding the new names. 

We have to discuss a bit about the orphan datasets. I would say we should probably delete them too.

File: packages/custom-storage/__tests__/SQLStorage.spec.ts (hunk 1 above)
Comment (neokore) [reply packages/custom-storage/__tests__/SQLStorage.spec.ts]: This was a quick fix for a mismatch between `lastModified` in object and `last_modified` in the DB; to avoid a deeper change when we wanted to finish v0.2 soon.
But thinking on it again, it could be better to create another field in columns for `nameDB` or just create an object adapter to transform between JS and DB.

File: packages/custom-storage/src/sql/SQLStorage.ts [hunk 2]
```diff
@@ -161,85 +164,19 @@ export class SQLStorage {
...
-        await this.shareDataset(tableName);
-      }
-    }
-
+    await this.uploadAndLinkDatasetsTo(insertedVis.id, datasets, overwriteDatasets, vis.isPrivate);
 
     return {
-      id,
-      lastModified,
+      ...insertedVis,
```
Comment (neokore) [inline packages/custom-storage/src/sql/SQLStorage.ts]: I think we need to change this order to avoid `...vis` to overwrite values from `insertedVis`

Comment (VictorVelarde) [reply packages/custom-storage/src/sql/SQLStorage.ts]: Do we? This is on creation, so 'id' and 'lastmodified' shouldn't be set on the 'vis' yet...

Comment (neokore) [reply packages/custom-storage/src/sql/SQLStorage.ts]: Sorry, I mean on update

File: packages/custom-storage/src/sql/SQLStorage.ts [hunk 3]
```diff
@@ -297,72 +234,18 @@ export class SQLStorage {
...
-      // We'll need some extra user info for this step, fetch this early on.
-      // CARTODBFY(...);
-      // END;
-    }
+    await this.uploadAndLinkDatasetsTo(updatedVis.id, datasets, true, vis.isPrivate);
 
     return {
-      id,
-      lastModified,
+      ...updatedVis,
```
Comment (neokore) [inline packages/custom-storage/src/sql/SQLStorage.ts]: Here is where we should swap variables

File: packages/custom-storage/__tests__/SQLStorage.spec.ts (hunk 1 above)
Comment (VictorVelarde) [reply packages/custom-storage/__tests__/SQLStorage.spec.ts]: Feel free to improve the 'FIELDS' with any mechanism you think is better to avoid mistakes

File: packages/custom-storage/src/sql/SQLStorage.ts (hunk 3 above)
Comment (VictorVelarde) [reply packages/custom-storage/src/sql/SQLStorage.ts]: Yes, it looks so. 👌

File: packages/custom-storage/src/sql/SQLStorage.ts [hunk 4]
```diff
@@ -350,6 +290,120 @@ export class SQLStorage {
...
+    RETURNING ${this.VIS_FIELDS.id.name}, ${this.VIS_FIELDS.lastModified.name}
+  `);
+
+    if (updatedResult.error) {
+     return null;
+   }
+
+    return {
+     id: vis.id,
+     lastModified: updatedResult.rows.length ? updatedResult.rows[0].lastmodified : vis.lastModified
```
Comment (VictorVelarde) [inline packages/custom-storage/src/sql/SQLStorage.ts]: I modified your `rows[0].lastModified` with `.lastmodified`. Double check it please

File: packages/custom-storage/src/sql/SQLStorage.ts (hunk 3 above)
Comment (neokore) [reply packages/custom-storage/src/sql/SQLStorage.ts]: After testing it, it appears that it keeps last modification field, so no changes needed here.

File: packages/custom-storage/src/sql/SQLStorage.ts (hunk 4 above)
Comment (neokore) [reply packages/custom-storage/src/sql/SQLStorage.ts]: It is right

**Comment by VictorVelarde** [comment]
Ready to merge after you can have a look at it

**Comment by VictorVelarde** [comment]
I have fixed the linter, merged master into and added the orphan datasets truncation (not tested yet)

File: packages/sql/src/RequestManager.ts [hunk 5]
```diff
@@ -127,7 +127,7 @@ export class RequestManager {
         const responseBody = await getResponseBody(response);
 
         const isTimeoutError = response.status === HTTP_ERRORS.TOO_MANY_REQUESTS &&
-          responseBody.detail === 'datasource';
+          (responseBody.detail === 'datasource' || responseBody.detail === 'rate-limit');
```
Comment (VictorVelarde) [inline packages/sql/src/RequestManager.ts]: ✨ Good catch!
